swarm/src/lib.rs

```diff
@@ -156,8 +166,8 @@ pub enum SwarmEvent<TBvEv> {
         /// Error that has been encountered.
         error: Box<dyn error::Error + Send>,
     },
-    /// Startng to try to reach the given peer.
-    StartConnect(PeerId),
+    /// A new dialing attempt has been initiated.
+    Dialing(PeerId, ConnectionId),
```

Comment:
There are two different ways to go for this event, in my opinion:

- Either it is emitted every single time we start an outgoing connection, in which case it should include the `Multiaddr` that we're trying to reach.
- Or we emit it when we start trying to reach the given peer (as was the case for `StartConnect`), in which case I don't think it should contain the `ConnectionId`.

Right now this `ConnectionId` refers to the first connection attempt (out of possibly many), and is not very useful.

Reply:
> Or we emit it when we start trying to reach the given peer (as was the case for StartConnect), in which case I don't think it should contain the ConnectionId.

Note that `StartConnect` was previously always emitted, even if `ExpandedSwarm::dial` didn't actually do anything. It is only now that it is only emitted when a new dialing attempt has indeed been started. I can see why you may not consider the connection ID currently useful, since the Swarm does not permit direct access to the underlying `Network` API (which I thought may be useful in the future, but is certainly not necessary now). However, `ConnectionEstablished` and `ConnectionClosed` also carry the `ConnectionId` and these are at least mandatory if we want to do https://github.com/libp2p/rust-libp2p/issues/1522. However, if you prefer for this PR we could remove all connection IDs from the `SwarmEvent`, if that sounds ok to you?

Reply:
> Note that StartConnect / Dialing was previously always emitted, even if ExpandedSwarm::dial didn't actually do anything.

That looks wrong to me.

> I can see why you may not consider the connection ID currently useful, since the Swarm does not permit direct access to the underlying Network API

Correct me if I'm wrong, but each dialing attempt gets its own `ConnectionId`, right?

Imagine if a certain node has two addresses A and B. When we start dialing address A, we emit a `Dialing` event with `A`'s `ConnectionId`. Supposing that dialing address A fails, we immediately start dialing B, which has a different `ConnectionId`, except that we don't report this new `ConnectionId` to the user in any way.

That's the problem that I was pointing out, but I'm not sure if I'm correct here.

Reply:
> > Note that StartConnect / Dialing was previously always emitted, even if ExpandedSwarm::dial didn't actually do anything.
> 
> That looks wrong to me.

I agree, which is why I changed it, but this has been the case [at least for a couple months](https://github.com/libp2p/rust-libp2p/blame/84b6a7d04ded33216d5579b629731988c721fb1f/swarm/src/lib.rs#L469-L470).

> > I can see why you may not consider the connection ID currently useful, since the Swarm does not permit direct access to the underlying Network API
> 
> Correct me if I'm wrong, but each dialing attempt gets its own `ConnectionId`, right?

A "dialing attempt" is by definition a sequence of "connection attempts" that terminates with the first successful connection or with the last failed connection, and each connection attempt gets its own `ConnectionId`, yes. There is no such thing as a "dialing attempt ID" (there never was).

> Imagine if a certain node has two addresses A and B. When we start dialing address A, we emit a `Dialing` event with `A`'s `ConnectionId`. Supposing that dialing address A fails, we immediately start dialing B, which has a different `ConnectionId`, except that we don't report this new `ConnectionId` to the user in any way.

That is correct. While the user is not actively notified though, all current dialing connections can be seen at any time in the `Peer` API of the `Network`. It is just the result of introducing the notion of `ConnectionId`s (and multiple connections in general) while preserving the existing dialing attempt semantics. We could certainly introduce another network event for each successive dialing attempt, but that seems out of scope of this PR.

For this PR, would you then prefer to just remove the `ConnectionId` from the `Dialing` event?

Reply:
> A "dialing attempt" is by definition a sequence of "connection attempts" to different addresses that terminates with the first successful connection or with the last failed connection

That was poor wording from me. By "dialing attempt" here I was thinking of connection attempt.

> For this PR, would you then prefer to just remove the ConnectionId from the Dialing event?

I would indeed prefer.

Reply:
Resolved by https://github.com/libp2p/rust-libp2p/pull/1519/commits/4fe71ed7166739602fe3760b7266c87ce92081ba. I also kept the connection IDs out of `SwarmEvent::ConnectionEstablished` and `SwarmEvent::ConnectionClosed` for now, since there is no point as long as the Swarm provides no APIs to work with connections directly.